README.md
Introduction
DataGrid

Getting Started
Import ag grid from the component library:

import {AgGrid} from “lai_webui”

Necessary Props:
rowData and columnData are required props. Pass your defined rows and columns to Data grid as below:

Example rows and columns: const columns = [ {field: “id”}, {field:” task”}, {field: “priority”} ];

const rows = [ {id: 1, task: “Task 1”, priority: “Low”}, {id: 2, task: “Task 2”, priority: “High”}, {id: 3, task: “Task 3”, priority: “Medium”}, {id: 4, task: “Task 4”, priority: “Low”}, ]

Pass them to rowData and columnData like below:

<AgGrid rowData={rows} columnData={columns} />

Props:
rowData: used to pass rows to Datagrid

columnData: used to pass columns to DataGrid

defaultColumnOptions: used to define default column options, it will apply all properties defined in it to all columns in the grid

columnReordering: used to enable reordering of columns

rowHeight : used to pass custom row height to the grid

headerRowHeight: used to pass custom header row height to the grid

summaryRowHeight: used to pass custom summary row height to the grid

direction: Used to set the direction in which to display the grid data direction={"ltr"} - to set direction left to right direction={"rtl"} - to set direction right to left. Grid default is ltr

className: Used to pass custom css className to the grid

rowClass: Used to pass custom css to to rows of the grid. (Will need to install and import css from linaria for using cell classnames) Example: const highlightClassname = css`

.rdg-cell {
    background-color: #9370db;
    color: white;
}`;
rowClass={(row) => row.id.includes("7") ? highlightClassname : undefined }

selectedCellHeaderStyle: To pass custom style object to header Example:

const selectedCellHeaderStyle = {
    backgroundColor: "red",
    fontSize: "12px",
};
selectedCellRowStyle: To pass custom style object for selected row Example:

const selectedCellRowStyle = {
    backgroundColor: "yellow",
};
selectedRows: To pass selected rows array (required for multiple row selection using checkbox)

onSelectedRowsChange: To pass function to be called on selected rows change (required for multiple row selection using checkbox)

onCopy: Accepts a function that gets called when a cell is copied

onPaste: Accepts a function that gets called when a cell is pasted

onScroll: Accepts a function that gets called user scrolls

onColumnResize: A function receiving column resize updates

restriction: Used to set restriction on copy and paste feature restriction prop to the AgGrid will allow you to enable or disable the copy/paste feature Example:

restriction={{
    copy: true,  // it should be not allowed table content copy
    paste: false, // it should be allow paste
}}
rowKeyGetter: A function returning a unique key/identifier per row. rowKeyGetter is required for row selection to work.

Example:

function rowKeyGetter(row) {
    return row.id;
}
rowKeyGetter={rowKeyGetter}
onRowsChange: A function receiving row updates. The first parameter is a new rows array with both the updated rows and the other untouched rows. The second parameter is an object with an indexes array highlighting which rows have changed by their index, and the column where the change happened.

Example:

const [rows, setRows] = useState(initialRows);

<AgGrid 
    columns={columns} 
    rows={rows} 
    onRowsChange={setRows} 
/>;
cellNavigationMode: Used to set the cell navigation mode It has folllowing options: "NONE" : No cell navigation "CHANGE_ROW": Goes over to the first cell of the next row after the last cell in current row "LOOP_OVER_ROW": Goes over to the first cell of the same row after the last cell The grid default is NONE

Example: cellNavigationMode={"CHANGE_ROW"} to change the cell navigation mode

SummaryRows: An optional array of summary rows, usually used to display total values for example. topSummaryRows: Used to set the top summary row/s

bottomSummaryRows: Used to set the bottom summary row/s

showSelectedRows: Used to show selected row count

export: To export grid to various formats pdf,csv,xlsx

groupBy: To pass array to perform row grouping on

rowGrouper: To pass function to perform row grouping

expandedGroupIds: To pass ids for group that should stay expanded

onExpandedGroupIdsChange: To pass function

enableVirtualization: to enable or disable virtualization of rows, it's enabled by default

frameworkComponents: to pass custom frameworkComponents to cells of the grid

style: Used to pass custom styling to the grid

serialNumber: to display serial number column in the grid

onCellClicked: Used to pass function to be performed on click of a cell Example: onCellClicked={(e)=> console.log(e)} // Will Print contents about the clicked cell and gridAPI and columnAPI actions to the console

onCellDoubleClicked: Used to pass function to be performed on double click of a cell Example:

onCellDoubleClicked={(e)=> console.log(e)} // Will Print contents about the double clicked cell and gridAPI and columnAPI actions to the console
onCellContextMenu: Used to pass function to be performed on right click of a cell Example:

onCellContextMenu={(e)=> console.log(e)} // Will Print contents about the right clicked cell and gridAPI and columnAPI actions to the console
onRowClicked: Used to pass function to be performed on click of a row Example:

onRowClicked={(e)=> console.log(e)} // Will Print contents about the clicked row and gridAPI and columnAPI actions to the console
onRowDoubleClicked: Used to pass function to be performed on double click of a row Example:

onRowDoubleClicked={(e)=> console.log(e)} // Will Print contents about the clicked row and gridAPI and columnAPI actions to the console
onGridReady: Used to pass function to be performed when grid is ready and will get the grid details and datas like gridApi,columnApi etc.. Example:

function gridReady(){
    alert("Grid is Ready to Show Details")
}

ongridReady={(props)=>{
    console.log(props);
    // Will Print contents about grid 
    gridReady()
    // then what ever functionalities we need to perform once grid 
    is ready those functionalities we need to pass here.
}}
Column Properties:
DataGrid provides following functionalities for column definitions:

const columns=[
{headerName: "PAXType", field: "PAXType", width: 160 },
{
    headerName: “Title”,
    field: “task”,
    width: 40, //To set column width
    maxWidth: 40, //To set column max width
    frozen: true, //To enable or disable given column as frozen
    filter: true,  // To enable or isable filtering for given column
    sortable:true,  //To enable or disable sorting for given column
    resizable: true, //To enable or disable resizing for the given column
    editable: false, //To enable or disable editing for given column
    cellEditor: textEditor, //To pass custom text editor component to the grid
    cellRenderer: () => <button> Save </button>, //To pass custom component in cell, can also pass it using  framework components
    validation:{
        style: { backgroundColor: "red", color: "blue" },
        method: (value) => value.slice(1) >= 100,
    }, //To pass validation function to perform on all cells in column
    formatter(props) {
        return <>{props.row.title}</>;
    }, //To pass custom formatter for every cell in column
    alignment: {type:"Date"}, //To align data in cells according to datatype. Available formats are : "Date","DateTime","Number"
}, ]

Sample Declaration:
<AgGrid
    columnData={columns}
    rowData={rows}
    rowKeyGetter={rowKeyGetter}
    onRowsChange={setRows}
    onCopy={handleCopy}
    onPaste={handlePaste}
    rowHeight={30} //To pass custom rowHeight
    serialNumber={true} // To display a serial number column in the grid
    showSelectedRows={true} // To display the count of selected rows
    selectedCellHeaderStyle={selectedCellHeaderStyle}
    selectedCellRowStyle={selectedCellRowStyle}
    selectedRows={selectedRows}
    onSelectedRowsChange={setSelectedRows}
    className="fill-grid"
    rowClass={(row) => row.id.includes("7") ? 
        highlightClassname : undefined 
    }
    direction={direction} 
    // To set the grid direction, default is ltr-left to right
    restriction={{
    copy: true,  // it should be not allowed table content copy
    paste: false, // it should be allow paste
    }}
/>
Features in datagrid:
Default Column Options Can pass default column options using the defaultColumnOptions prop

Example:
const default={ resizable: true, sortable: true }
<AgGrid defaultColumnOption={default}>
Serial Number: To have a serial number column in your grid, pass serialNumber = {true} or serialNumber prop to the AgGrid declaration.

Cell validation: We can be able to high light cell data is valid or not for column vise. For example we are setting one column data is valid for more than 100 only means, it will high light background color will be red in which cell of column is below 100.

const columns= [{
field:” age”,
headerName:” Age”,
validation: {
        style: { backgroundColor: "red", color: "blue" },  
        // not valid cell style ,suppose not giving style it will set only default background color red.
        method: (value) => value >= 18, // validation method
    },
}]
Alignment according to DataTypes:

Based on data type (number, text, date, time, datetime, money, etc.) column vise align (start, center, end) data is possible for column definition area.

Auto detect data type and align

While setting alignment: true will detect automatically data type and alignment method.

const columns= [{
    headerName:” Name”,
    field:” name”,
    alignment: true
}]

Example Data Types and default align method:

“Victor” =text   // start

12345 = number // end

“₹100” = currency // end      maximum auto detectable currencies  £, $, ₹, €, ¥, ₣, ¢

Mar/23/2020, 30-02-20220, 2020/27/April = date // end

Mar/23/2020 19:40:23, 30-02-20220 09:20:55, 2020/27/April 01:46:12 am = datetime // end

Not able to detect perfect data type to set as a center.
Manually set the Type and alignment method:

To set the property like this alignment: {type:” text”, align:” end”}. Suppose not setting alignment method it will take default method.
Framework Components: You can also pass custom framework components to the cellRenderer in a column definition.

const customComponents={
    Btn:() => <Button> Save </Button> // Return your required component in frameworkComponents definition
}

In the column definition, pass "Btn" as string to the cellRenderer:

const columns=[
    {field: "id" , headerName:"ID"},
    {field: "save", cellRenderer: "Btn"}, //Will render the button in every cell of given column
]

Pass customComponents to the frameworkComponents prop in the AgGrid declaration

<AgGrid
    rowData={rows}
    columnData={columns}
    frameworkComponents={customComponents}
/>
Export to various formats pdf, csv, xlsx:

We can export the table data into file formats like pdf, csv and xlsx(excel), it should be achievable for accessing export property. AgGird will show buttons automatically in table top right corner. AgGird provides a separate button for each format.

Example:

<AgGrid  
    columnData={columns}
    rowData={rows}
    export= {{
      pdfFileName: "TableData",  //  To export pdf file
      csvFileName: "TableData", // To export csv file
      excelFileName:"TableData", // To export excel or xlsx file
    }}
/>
Row selection using checkbox Pass selection, onSelecteRowsChange and rowKeyGetter function for multiple row selection using checkbox

Example:

const [selectedRows,setSelectedRows]=useState([]);

function rowKeyGetter(row) {
    return row.id;
}
<AgGrid
    columnData={columns}
    rowData={rows}
    selection={true}
    onSelectedRowsChange={setSelectedRows}
    rowKeyGetter={rowKeyGetter}
/>
Row Selection without checkbox Pass rowKeyGetter function, and rowSelection prop. There are two modes of row selection, "single" and "multiple". Single: Can only select one row Multiple: Can select multiple rows

Example:

function rowKeyGetter(row) {
    return row.id;
}
id should be unique
<AgGrid
    columnData={columns}
    rowData={rows}
    rowKeyGetter={rowKeyGetter}
    rowSelection={"multiple"} //Can select multiple rows
/>
Status Bar: Shows selected row count. Pass prop showSelectedRows to view status bar Example:

<AgGrid
    columnData={columns}
    rowData={rows}
    showSelectedRows={true}
/>
Filtering: To enable filtering for any column, pass filter: true in that column's definition

Example:
  const columns= [{
    headerName:” Name”,
    field:” name”,
    filter: true //to enable filtering
}]
Sorting To enable sorting for any column, pass sortable: true in that column's definition

Example:
  const columns= [{
    headerName:” Name”,
    field:” name”,
    sortable: true //to enable sorting
}]
Column Spanning:

To merge column vise two are more cells is possible to pass the function of colSpan in column defintion area. And also possible to Span particular rows only.

Example:

  const columns= [{
    headerName:” Name”,
    field:” name”,
    colSpan: (params) => {
    // params will give the type of row
    // normal row cell in second row only
     if (params.type === "ROW" &&
        params.rowIndex === 1) {
       return 2;
    // need to retrun number of ells     going to merge/span
     }
    // header row cell
     if (params.type === "HEADER") {
       return 2;
     }
   },
},
{
    headerName:” Department”,
    field:” department”,
}
]
Copy/Paste with or without restrictions : Some time table will contain sensitive data’s ,so need to stop easy cloning it should be possible to set restriction property. In that case, passing the restriction prop to the AgGrid will allow you to enable or disable the copy/paste feature

restriction={{
    copy: true,  // it should be not allowed table content copy
    paste: false, // it should be allow paste
  }}
Pagination: Pagination allows the grid to paginate rows, removing the need for a vertical scroll to view more data and pagination based better view and user experience. Example:

<AgGrid
    columnData={columns}
    rowData={rows}
    pagination={true}
    // used to enable Grid Pagination
    defaultPage={3}
    // used to set default showing page  
    paginationAutoPageSize={true} 
    // used to set  number of rows perpage to auto 
    paginationPageSize={20}
    // used to set number of rows perpage 
    suppressPaginationPanel={true}
    // used to hide the pagination panel but pagination was enabled and
    will control some other way means with help of gridApi  
/>
Context Menu : You can customise the context menu by providing a getContextMenuItems() callback. Each time the context menu is to be shown, the callback is called to retrieve the menu items.

Example:

const getContextMenuItems = useCallback(() => {
let menu = [
{  
    name: "Alert ", // Option Name

Changelog.md
### **20230616.3**
Build Number: 17743 
#### Associated Work Items (1)
  1. Task 40164: React data Grid - integration to library
### **20230616.2**
Build Number: 17739 
#### Associated Work Items (1)
  1. Task 40164: React data Grid - integration to library
### **20230616.1**
Build Number: 17707 
#### Associated Work Items (8)
  1. Product Backlog Item 28428: React Data Grid - Grid API Functionality
  1. Task 28118: Grid API -  forEachLeafNode() - Code Review
  1. Task 28120: Grid API - getSelectedNodes() - Code Review
  1. Task 28152: Grid API - getSelectedNodes() - Integrate
  1. Task 28158: Grid API - applyTransaction() - Code Review
  1. Task 28469: React Data Grid - Select column and selected count like (3 of 10): Unit Testing
  1. Task 31231: React Data Grid - API Functionality - Code Review
  1. Task 31236: React Data Grid - RowNode: Code Review


src/demos/NoRows.js
// import { createContext, useMemo, useState } from "react";

// import DataGrid from "../components/datagrid/DataGrid";

// import { SerialNumberColumn } from "../components/datagrid/Columns";

// function EmptyRowsRenderer() {
//   return (
//     <div style={{ textAlign: "center", gridColumn: "1/-1" }}>
//       Nothing to show
//       <span lang="ja" title="ショボーン">
//         (´・ω・`)
//       </span>
//     </div>
//   );
// }


// const FilterContext = createContext(undefined);

// function rowKeyGetter(row) {
//   return row.id;
// }

// export default function NoRows({ direction }) {

//   const [uploadfiles, setuploadFiles] = useState([]);
//   const [file, setFile] = useState([])
// console.log('uploadfiles', uploadfiles)
//   const handleDelete = (key) => {
//     console.log('uploadfiles', uploadfiles)
//     console.log('file', file)
//     uploadfiles.splice(key.rowIndex, 1);
//     file.splice(key.rowIndex, 1);
//     setuploadFiles([...uploadfiles]);
//     setFile([...file]);
//     console.log("delelte ", file);
//     console.log('key', key)
//   };
//   const frameworkComponents = {
//     CheckBox: (props) => (
//       <button style={{ width: "100%" }} onClick={()=>{
//         return handleDelete(props);
//       }}>
//         Delete
//       </button>
//     ),
//   };


//   const [selectedRows, onSelectedRowsChange] = useState(() => new Set());
//   const [filters, setFilters] = useState({
//     title: "",
//     count: "",
//     enabled: true,
//   });

//   const columns = useMemo(() => {
//     return [
//       SerialNumberColumn,
//       {
//         field: "id",
//         headerName: "ID",
//         width: 50,
//         haveChildren: false,
//         topHeader: "id",
//         cellWidth: 60,
//         frozen: true,
//       },
//       {
//         field: "rdrd",
//         headerName: "AASS",
//         haveChildren: false,
//         topHeader: "rdrd",
//         cellWidth: 60,
//         cellRenderer: "CheckBox",
//       },

//       {
//         field: "title",
//         headerName: "Title",
//         haveChildren: true,
//         // frozen: true,

//         children: [
//           // SelectColumn,
//           {
//             field: "aaaa",
//             headerName: "AAAA",

//             haveChildren: true,
//             children: [
//               {
//                 field: "vvvv",
//                 headerName: "VVVV",
//                 haveChildren: false,
//                 cellWidth: 100,
//                 topHeader: "title",
//               },

//               {
//                 field: "rrrr",
//                 headerName: "RRRR",
//                 haveChildren: false,
//                 cellWidth: 100,
//                 topHeader: "title",
//               },
//               {
//                 field: "uuuu",
//                 headerName: "UUUU",
//                 haveChildren: false,
//                 cellWidth: 100,
//                 topHeader: "title",
//               },
//             ],
//           },
//           {
//             field: "bbbb",
//             headerName: "BBBB",
//             haveChildren: true,
//             children: [
//               {
//                 field: "wsdc",
//                 headerName: "HGTF",
//                 haveChildren: false,
//                 cellWidth: 100,
//                 topHeader: "title",
//               },
//               {
//                 field: "yugd",
//                 headerName: "HGFBGV",
//                 haveChildren: false,
//                 cellWidth: 100,
//                 topHeader: "title",
//               },
//             ],
//           },
//           {
//             field: "cccc",
//             headerName: "CCCC",
//             cellWidth: 100,
//             haveChildren: true,
//             children: [
//               {
//                 field: "yugd",
//                 headerName: "HGFBGV",
//                 haveChildren: false,
//                 cellWidth: 100,
//                 topHeader: "title",
//               },
//               {
//                 field: "yugd",
//                 headerName: "HGFBGV",
//                 haveChildren: true,
//                 topHeader: "title",
//                 cellWidth: 100,
//                 children: [
//                   {
//                     field: "cvdcv",
//                     headerName: "FGHT",
//                     haveChildren: false,
//                     topHeader: "title",
//                     cellWidth: 60,
//                   },
//                   {
//                     field: "cvacv",
//                     headerName: "FGHT",
//                     haveChildren: false,
//                     topHeader: "title",
//                     cellWidth: 60,
//                   },
//                 ],
//               },
//             ],
//           },
//         ],
//       },
//       {
//         field: "cvcv",
//         headerName: "FGHT",
//         haveChildren: false,
//         topHeader: "cvcv",
//         cellWidth: 60,
//       },
//       {
//         field: "erer",
//         headerName: "FGHT",
//         haveChildren: false,
//         topHeader: "erer",
//         cellWidth: 60,
//       },
//       {
//         field: "count",
//         headerName: "Count",
//         haveChildren: true,
//         children: [
//           // SelectColumn,
//           {
//             field: "nnnn",
//             headerName: "NNNN",
//             haveChildren: true,
//             children: [
//               {
//                 field: "xxxx",
//                 headerName: "XXXX",
//                 haveChildren: false,
//                 cellWidth: 100,
//                 topHeader: "count",
//               },
//               {
//                 field: "jjjj",
//                 headerName: "JJJJ",
//                 haveChildren: true,
//                 children: [
//                   {
//                     field: "ffff",
//                     headerName: "FFFF",
//                     haveChildren: false,
//                     cellWidth: 100,
//                     topHeader: "count",
//                   },
//                   {
//                     field: "vvvv",
//                     headerName: "VVVV",
//                     haveChildren: true,
//                     children: [
//                       {
//                         field: "llll",
//                         headerName: "LLLL",
//                         haveChildren: false,
//                         cellWidth: 100,
//                         topHeader: "count",
//                       },
//                       {
//                         field: "pppp",
//                         headerName: "PPPP",
//                         haveChildren: true,
//                         children: [
//                           {
//                             field: "eeee",
//                             headerName: "EEEE",
//                             haveChildren: false,
//                             cellWidth: 100,
//                             topHeader: "count",
//                           },
//                           {
//                             field: "pppp",
//                             headerName: "PPPP",
//                             haveChildren: true,
//                             cellWidth: 100,
//                             topHeader: "count",
//                             children: [
//                               {
//                                 field: "eeee",
//                                 headerName: "EEEE",
//                                 haveChildren: false,
//                                 cellWidth: 100,
//                                 topHeader: "count",
//                               },
//                               {
//                                 field: "pppp",
//                                 headerName: "PPPP",
//                                 haveChildren: false,
//                                 cellWidth: 100,
//                                 topHeader: "count",
//                               },
//                             ],
//                           },
//                         ],
//                       },
//                     ],
//                   },
//                 ],
//               },
//             ],
//           },
//           {
//             field: "oooo",
//             headerName: "OOOO",
//             cellWidth: 100,
//             haveChildren: false,
//             topHeader: "count",
//           },
//           {
//             field: "qqqq",
//             headerName: "QQQQ",
//             cellWidth: 100,
//             haveChildren: false,
//             topHeader: "count",
//           },
//         ],
//       },
//     ];
//   }, []);

//   const rows = [
//     {
//       id: 3,
//       oooo: "JGRF",
//       aaaa: "wsws",
//       vvvv: "Ssss",
//       nnnn: "gege",
//       qqqq: "kjkj",
//       llll: "llll",
//       rrrrr: "thgf",
//       uuuu: "iugy",
//       bbbb: "vfvf",
//       cccc: "sdsd",
//       xxxx: "sdsa",
//       ffff: "aqaq",
//       eeee: "tyty",
//       pppp: "frfr",
//     },
//     {
//       id: 4,
//       vvvv: "assdd",
//       nnnn: "gege",
//       qqqq: "kjkj",
//       llll: "llll",
//       rrrrr: "thgf",
//       uuuu: "iugy",
//       bbbb: "vfvf",
//       cccc: "sdsd",
//       xxxx: "sdsa",
//       ffff: "aqaq",
//       eeee: "tyty",
//       pppp: "frfr",
//     },
//     {
//       id: 5,
//       vvvv: "Ssss",
//       nnnn: "gege",
//       qqqq: "kjkj",
//       llll: "llll",
//       rrrrr: "thgf",
//       uuuu: "iugy",
//       bbbb: "vfvf",
//       cccc: "sdsd",
//       xxxx: "sdsa",
//       ffff: "aqaq",
//       eeee: "tyty",
//       pppp: "frfr",
//     },
//     {
//       id: 6,
//       vvvv: "Ssss",
//       nnnn: "gege",
//       qqqq: "kjkj",
//       llll: "llll",
//       rrrr: "wrerer",
//       rrrrr: "thgf",
//       uuuu: "iugy",
//       bbbb: "vfvf",
//       cccc: "sdsd",
//       xxxx: "sdsa",
//       ffff: "aqaq",
//       eeee: "tyty",
//       pppp: "frfr",
//     },
//     {
//       id: 7,
//       vvvv: "Ssss",
//       nnnn: "gege",
//       qqqq: "kjkj",
//       llll: "llll",
//       rrrrr: "thgf",
//       uuuu: "iugy",
//       bbbb: "vfvf",
//       cccc: "sdsd",
//       xxxx: "sdsa",
//       ffff: "aqaq",
//       eeee: "tyty",
//       pppp: "frfr",
//     },
//     {
//       id: 8,
//       vvvv: "Ssss",
//       nnnn: "gege",
//       qqqq: "kjkj",
//       llll: "llll",
//       rrrrr: "thgf",
//       uuuu: "iugy",
//       bbbb: "vfvf",
//       cccc: "sdsd",
//       xxxx: "sdsa",
//       ffff: "aqaq",
//       eeee: "tyty",
//       pppp: "frfr",
//     },
//     {
//       id: 9,
//       vvvv: "Ssss",
//       nnnn: "gege",
//       qqqq: "kjkj",
//       llll: "llll",
//       rrrrr: "thgf",
//       uuuu: "iugy",
//       bbbb: "vfvf",
//       cccc: "sdsd",
//       xxxx: "sdsa",
//       ffff: "aqaq",
//       eeee: "tyty",
//       pppp: "frfr",
//     },
//     {
//       id: 10,
//       vvvv: "Ssss",
//       nnnn: "gege",
//       qqqq: "kjkj",
//       llll: "llll",
//       rrrrr: "thgf",
//       uuuu: "iugy",
//       bbbb: "vfvf",
//       cccc: "sdsd",
//       xxxx: "sdsa",
//       ffff: "aqaq",
//       eeee: "tyty",
//       pppp: "frfr",
//     },
//     {
//       id: 11,
//       vvvv: "Ssss",
//       nnnn: "gege",
//       qqqq: "kjkj",
//       llll: "llll",
//       rrrrr: "thgf",
//       uuuu: "iugy",
//       bbbb: "vfvf",
//       cccc: "sdsd",
//       xxxx: "sdsa",
//       ffff: "aqaq",
//       eeee: "tyty",
//       pppp: "frfr",
//     },
//     {
//       id: 12,
//       vvvv: "Ssss",
//       nnnn: "gege",
//       qqqq: "kjkj",
//       llll: "llll",
//       rrrrr: "thgf",
//       uuuu: "iugy",
//       bbbb: "vfvf",
//       cccc: "sdsd",
//       xxxx: "sdsa",
//       ffff: "aqaq",
//       eeee: "tyty",
//       pppp: "frfr",
//     },
//     {
//       id: 13,
//       vvvv: "Ssss",
//       nnnn: "gege",
//       qqqq: "kjkj",
//       llll: "llll",
//       rrrrr: "thgf",
//       uuuu: "iugy",
//       bbbb: "vfvf",
//       cccc: "sdsd",
//       xxxx: "sdsa",
//       ffff: "aqaq",
//       eeee: "tyty",
//       pppp: "frfr",
//     },
//     {
//       id: 14,
//       vvvv: "Ssss",
//       nnnn: "gege",
//       qqqq: "kjkj",
//       llll: "llll",
//       rrrrr: "thgf",
//       uuuu: "iugy",
//       bbbb: "vfvf",
//       cccc: "sdsd",
//       xxxx: "sdsa",
//       ffff: "aqaq",
//       eeee: "tyty",
//       pppp: "frfr",
//     },
//     {
//       id: 15,
//       vvvv: "Ssss",
//       nnnn: "gege",
//       ccrcc: "dfhhgh",
//       qqqq: "kjkj",
//       llll: "llll",
//       rrrrr: "thgf",
//       uuuu: "iugy",
//       bbbb: "vfvf",
//       cccc: "sdsd",
//       xxxx: "sdsa",
//       ffff: "aqaq",
//       eeee: "tyty",
//       pppp: "frfr",
//     },
//     {
//       id: 16,
//       vvvv: "Ssss",
//       nnnn: "gege",
//       qqqq: "kjkj",
//       llll: "llll",
//       rrrrr: "thgf",
//       uuuu: "iugy",
//       bbbb: "vfvf",
//       cccc: "sdsd",
//       xxxx: "sdsa",
//       ffff: "aqaq",
//       eeee: "tyty",
//       pppp: "frfr",
//     },
//     {
//       id: 17,
//       vvvv: "Ssss",
//       nnnn: "gege",
//       qqqq: "kjkj",
//       llll: "llll",
//       rrrrr: "thgf",
//       uuuu: "iugy",
//       bbbb: "vfvf",
//       cccc: "sdsd",
//       xxxx: "sdsa",
//       ffff: "aqaq",
//       eeee: "tyty",
//       pppp: "frfr",
//     },
//     {
//       id: 18,
//       vvvv: "Ssss",
//       nnnn: "gege",
//       qqqq: "kjkj",
//       llll: "llll",
//       rrrrr: "thgf",
//       uuuu: "iugy",
//       bbbb: "vfvf",
//       cccc: "sdsd",
//       xxxx: "sdsa",
//       ffff: "aqaq",
//       eeee: "tyty",
//       pppp: "frfr",
//     },
//     {
//       id: 19,
//       vvvv: "Ssss",
//       nnnn: "gege",
//       qqqq: "kjkj",
//       llll: "llll",
//       rrrrr: "thgf",
//       uuuu: "iugy",
//       bbbb: "vfvf",
//       cccc: "sdsd",
//       xxxx: "sdsa",
//       ffff: "aqaq",
//       eeee: "tyty",
//       pppp: "frfr",
//     },
//     {
//       id: 20,
//       vvvv: "Ssss",
//       nnnn: "gege",
//       qqqq: "kjkj",
//       llll: "llll",
//       rrrrr: "thgf",
//       uuuu: "iugy",
//       bbbb: "vfvf",
//       cccc: "sdsd",
//       xxxx: "sdsa",
//       ffff: "aqaq",
//       eeee: "tyty",
//       pppp: "frfr",
//     },
//     {
//       id: 21,
//       vvvv: "Ssss",
//       nnnn: "gege",
//       qqqq: "kjkj",
//       llll: "llll",
//       rrrrr: "thgf",
//       uuuu: "iugy",
//       bbbb: "vfvf",
//       cccc: "sdsd",
//       xxxx: "sdsa",
//       ffff: "aqaq",
//       eeee: "tyty",
//       pppp: "frfr",
//     },
//     {
//       id: 22,
//       vvvv: "Ssss",
//       nnnn: "gege",
//       qqqq: "kjkj",
//       llll: "llll",
//       rrrrr: "thgf",
//       uuuu: "iugy",
//       bbbb: "vfvf",
//       cccc: "sdsd",
//       xxxx: "sdsa",
//       ffff: "aqaq",
//       eeee: "tyty",
//       pppp: "frfr",
//     },
//     {
//       id: 23,
//       vvvv: "Ssss",
//       nnnn: "gege",
//       qqqq: "kjkj",
//       llll: "llll",
//       rrrrr: "thgf",
//       uuuu: "iugy",
//       bbbb: "vfvf",
//       cccc: "sdsd",
//       xxxx: "sdsa",
//       ffff: "aqaq",
//       eeee: "tyty",
//       pppp: "frfr",
//     },
//     {
//       id: 24,
//       vvvv: "Ssss",
//       nnnn: "gege",
//       qqqq: "kjkj",
//       llll: "llll",
//       rrrrr: "thgf",
//       uuuu: "iugy",
//       bbbb: "vfvf",
//       cccc: "sdsd",
//       xxxx: "sdsa",
//       ffff: "aqaq",
//       eeee: "tyty",
//       pppp: "frfr",
//     },
//     {
//       id: 25,
//       vvvv: "Ssss",
//       nnnn: "gege",
//       qqqq: "kjkj",
//       llll: "llll",
//       rrrrr: "thgf",
//       uuuu: "iugy",
//       bbbb: "vfvf",
//       cccc: "sdsd",
//       xxxx: "sdsa",
//       ffff: "aqaq",
//       eeee: "tyty",
//       pppp: "frfr",
//     },
//     {
//       id: 26,
//       vvvv: "Ssss",
//       nnnn: "gege",
//       qqqq: "kjkj",
//       llll: "llll",
//       rrrrr: "thgf",
//       uuuu: "iugy",
//       bbbb: "vfvf",
//       cccc: "sdsd",
//       xxxx: "sdsa",
//       ffff: "aqaq",
//       eeee: "tyty",
//       pppp: "frfr",
//     },
//     {
//       id: 27,
//       vvvv: "Ssss",
//       nnnn: "gege",
//       qqqq: "kjkj",
//       llll: "llll",
//       rrrrr: "thgf",
//       uuuu: "iugy",
//       bbbb: "vfvf",
//       cccc: "sdsd",
//       xxxx: "sdsa",
//       ffff: "aqaq",
//       eeee: "tyty",
//       pppp: "frfr",
//     },
//     {
//       id: 28,
//       vvvv: "Ssss",
//       nnnn: "gege",
//       qqqq: "kjkj",
//       llll: "llll",
//       rrrrr: "thgf",
//       uuuu: "iugy",
//       bbbb: "vfvf",
//       cccc: "sdsd",
//       xxxx: "sdsa",
//       ffff: "aqaq",
//       eeee: "tyty",
//       pppp: "frfr",
//     },
//     {
//       id: 29,
//       vvvv: "Ssss",
//       nnnn: "gege",
//       qqqq: "kjkj",
//       llll: "llll",
//       rrrrr: "thgf",
//       uuuu: "iugy",
//       bbbb: "vfvf",
//       cccc: "sdsd",
//       xxxx: "sdsa",
//       ffff: "aqaq",
//       eeee: "tyty",
//       pppp: "frfr",
//     },
//     {
//       id: 30,
//       vvvv: "Ssss",
//       nnnn: "gege",
//       qqqq: "kjkj",
//       llll: "llll",
//       rrrrr: "thgf",
//       uuuu: "iugy",
//       bbbb: "vfvf",
//       cccc: "sdsd",
//       xxxx: "sdsa",
//       ffff: "aqaq",
//       eeee: "tyty",
//       pppp: "frfr",
//     },
//     {
//       id: 31,
//       vvvv: "Ssss",
//       nnnn: "gege",
//       qqqq: "kjkj",
//       llll: "llll",
//       rrrrr: "thgf",
//       uuuu: "iugy",
//       bbbb: "vfvf",
//       cccc: "sdsd",
//       xxxx: "sdsa",
//       ffff: "aqaq",
//       eeee: "tyty",
//       pppp: "frfr",
//     },
//     {
//       id: 32,
//       vvvv: "Ssss",
//       nnnn: "gege",
//       qqqq: "kjkj",
//       llll: "llll",
//       rrrrr: "thgf",
//       uuuu: "iugy",
//       bbbb: "vfvf",
//       cccc: "sdsd",
//       xxxx: "sdsa",
//       ffff: "aqaq",
//       eeee: "tyty",
//       pppp: "frfr",
//     },
//     {
//       id: 33,
//       vvvv: "Ssss",
//       nnnn: "gege",
//       qqqq: "kjkj",
//       llll: "llll",
//       rrrrr: "thgf",
//       uuuu: "iugy",
//       bbbb: "vfvf",
//       cccc: "sdsd",
//       xxxx: "sdsa",
//       ffff: "aqaq",
//       eeee: "tyty",
//       pppp: "frfr",
//     },
//     {
//       id: 34,
//       vvvv: "Ssss",
//       nnnn: "gege",
//       qqqq: "kjkj",
//       llll: "llll",
//       rrrrr: "thgf",
//       uuuu: "iugy",
//       bbbb: "vfvf",
//       cccc: "sdsd",
//       xxxx: "sdsa",
//       ffff: "aqaq",
//       eeee: "tyty",
//       pppp: "frfr",
//     },
//     {
//       id: 35,
//       vvvv: "Ssss",
//       nnnn: "gege",
//       qqqq: "kjkj",
//       llll: "llll",
//       rrrrr: "thgf",
//       uuuu: "iugy",
//       bbbb: "vfvf",
//       cccc: "sdsd",
//       xxxx: "sdsa",
//       ffff: "aqaq",
//       eeee: "tyty",
//       pppp: "frfr",
//     },
//     {
//       id: 36,
//       vvvv: "Ssss",
//       nnnn: "gege",
//       qqqq: "kjkj",
//       llll: "llll",
//       rrrrr: "thgf",
//       uuuu: "iugy",
//       bbbb: "vfvf",
//       cccc: "sdsd",
//       xxxx: "sdsa",
//       ffff: "aqaq",
//       eeee: "tyty",
//       pppp: "frfr",
//     },
//   ];

//   return (
//     <FilterContext.Provider value={filters}>
//       <DataGrid
//         columnData={columns}
//         rowData={rows}
//         renderers={{ noRowsFallback: <EmptyRowsRenderer /> }}
//         selectedRows={selectedRows}
//         onSelectedRowsChange={onSelectedRowsChange}
//         frameworkComponents={frameworkComponents}
//         headerRowHeight={24}
//         rowKeyGetter={rowKeyGetter}
//         className="fill-grid"
//         // direction={direction}
//       />
//     </FilterContext.Provider>
//   );
// }
